src/ucp/core/ucp_request.inl

```diff
@@ -384,18 +384,24 @@ ucp_send_request_add_reg_lane(ucp_request_t *req, ucp_lane_index_t lane)
 
     if (ucs_likely(UCP_DT_IS_CONTIG(req->send.datatype))) {
         md_map = req->send.state.dt.dt.contig.md_map;
-    } else if (req->send.state.dt.dt.iov.dt_reg != NULL)  {
-        ucs_assert(UCP_DT_IS_IOV(req->send.datatype));
+    } else if (UCP_DT_IS_IOV(req->send.datatype) &&
+               (req->send.state.dt.dt.iov.dt_reg != NULL)) {
         /* dt_reg can be NULL if underlying UCT TL doesn't require
          * memory handle for for local AM/GET/PUT operations
          * (i.e. UCT_MD_FLAG_NEED_MEMH is not set) */
         /* Can use the first DT registration element, since
          * they have the same MD maps */
         md_map = req->send.state.dt.dt.iov.dt_reg[0].md_map;
     } else {
+        ucs_assert(!UCP_DT_IS_CONTIG(req->send.datatype) &&
+                   (!UCP_DT_IS_IOV(req->send.datatype) ||
+                    ((req->send.state.dt.dt.iov.dt_reg == NULL) &&
+                     (req->send.state.dt.dt.iov.iov_offset == 0) &&
+                     (req->send.state.dt.dt.iov.iovcnt == 0))));
```

Comment:
not sure we need this assertion, isn't it already part of the 'if' condition in lines 387-388?

Reply:
yes, it could be removed - done